[#19] 숙명포털 로그인 crawler

kohseongyeon opened the pull request:
Jsoup을 이용하여 숙명포털에 로그인하는 crawler를 구현하였습니다.

- 로그인 페이지에 접속하여, 로그인 처리 요청 시에 함께 전송해야 하는 토큰 가져오기
- 로그인 폼 데이터와 토큰을 Jsoup connection의 데이터로 추가하여, 로그인 요청
- 로그인 요청 Connection response에서 얻어지는 `JSESSIONID` 유무에 따라, 로그인 성공 여부 판별

kohseongyeon committed:
[#19] 숙명포털 로그인 crawler

kohseongyeon commented on authentication-api/src/main/java/com/bluebird/pipit/portal/PortalLoginCrawler.java:
```diff
@@ -0,0 +1,65 @@
+
+		Map<String, String> loginPageCookie = loginPageResponse.cookies();
+		Document loginPageDocument = loginPageResponse.parse();
+		String j_salt = loginPageDocument.select("input.j_salt").val();
```
로그인 처리를 위해서는, 로그인 페이지에 접속해야만 얻어낼 수 있는 토큰을 함께 전송해야 합니다.
숙명포털 로그인 페이지의 Request Header에 따르면 `j_salt` 라는 이름의 토큰을 로그인 요청 시 함께 전송합니다.

따라서 로그인 페이지의 response 객체를 통해 `j_salt`의 값을 얻어옵니다.

kohseongyeon commented on authentication-api/src/main/java/com/bluebird/pipit/portal/PortalLoginCrawler.java:
```diff
@@ -0,0 +1,65 @@
+		Map<String, String> loginCookie = response.cookies();
+
+		return loginCookie.containsKey("JSESSIONMARKID");
+	}
```
로그인 처리 요청 후에 얻어지는 쿠키입니다. 

로그인 성공 시, `JSESSIONMARKID` 라는 key값으로 세션 ID가 생성됩니다.
로그인 실패 시, 세션ID는 생성되지 않습니다.

juhyeonme replied:
`JSESSIONMARKID` 라는 키 값은 바뀌지 않는 값이지요? 
그렇다면 클래스 상단부에 `private static final` 선언해주는 게 좋을 것 같아요

juhyeonme commented on authentication-api/build.gradle:
```diff
@@ -34,6 +34,8 @@ dependencies {
     testCompileOnly("org.projectlombok:lombok")
     testAnnotationProcessor("org.projectlombok:lombok")
+
+    compile ('org.jsoup:jsoup:1.8.3')
```
@kohseongyeon 
생각해볼만한 점이 세가지 생각납니다.

1. `compile` 키워드는 deprecated 되었어요. 
  `api` 로 수정해주시면 좋을 것 같아요

2. `jsoup` dependency 는 테스트가 아닌 비즈니스 로직에서 사용되는 라이브러리인 만큼,
    `dependencies {...}` 내에서 순서있게 정의되면 좋을 것 같아요. 
    테스트에서 사용되는 dependency 가 아래쪽, 나머지는 위쪽에 위치하면 깔끔할 것 같습니다.

```java
    dependencies {
        implementation("org.springframework.boot:spring-boot-starter-web")

        api("org.jsoup:jsoup:1.8.3")

        compileOnly("org.projectlombok:lombok")
        annotationProcessor("org.projectlombok:lombok")

        testImplementation("org.springframework.boot:spring-boot-starter-test")

        testCompileOnly("org.projectlombok:lombok")
        testAnnotationProcessor("org.projectlombok:lombok")
    }
```

3. 지금 서버쪽에선 `spring-dependency-manager` 의 도움을 받아, 스프링 버전에 최적화된 버전을 자동으로 셋팅해주는 설정을 사용하고 있어요 (#2)
  jsoup 도 이걸 활용할 수 있는지 알아보고, 
spring dependency 의 관리를 받지 못하는 라이브러리라면 `ext` 를 정의해서 버전을 따로 관리해주어도 편리할 것 같아요

```groovy
ext {
   JSOUP_VERSION = "1.8.3"
}

...

dependencies {
    api("org.jsoup:jsoup:${JSOUP_VERSION}")
}
```

juhyeonme commented on authentication-api/src/main/java/com/bluebird/pipit/portal/PortalLoginCrawler.java:
```diff
@@ -0,0 +1,65 @@
+		loginFormData.put("j_password", password);
+		loginFormData.put("saveid", "N");
+
+		Connection.Response response = Jsoup.connect("https://portal.sookmyung.ac.kr/irj/portal")
+			.userAgent(userAgent)
+			.timeout(3000)
+			.header("Origin", "https://portal.sookmyung.ac.kr")
+			.header("Referer", "https://portal.sookmyung.ac.kr/irj/portal")
+			.header("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9")
+			.header("Content-Type", "application/x-www-form-urlencoded")
+			.header("Accept-Encoding", "gzip, deflate, br")
+			.header("Accept-Language", "ko-KR,ko;q=0.9,en-US;q=0.8,en;q=0.7")
```
여기에 사용된 값들은 다음 프로세스로 가져와서, 따로 관리해주는 편이 나을 것 같아요
- `application.yml` 파일에 등록 -> `@ConfigurationProperties` class -> 여기 클래스에서 `get`

juhyeonme commented on authentication-api/src/main/java/com/bluebird/pipit/portal/PortalLoginCrawler.java:
```diff
@@ -0,0 +1,65 @@
+
+		Map<String, String> loginCookie = response.cookies();
+
+		return loginCookie.containsKey("JSESSIONMARKID");
```
`response.cookies()` 를 `loginCookie` 라는 변수에 담았던 이유가 있었을까요?
이렇게 한번에 return 해버리면 어떨까 싶습니다.

```suggestion
		return response.cookies()
		    .containsKey("JSESSIONMARKID");
```

juhyeonme commented on authentication-api/src/main/java/com/bluebird/pipit/portal/PortalLoginCrawler.java:
```diff
@@ -0,0 +1,65 @@
+import java.util.Map;
+
+public class PortalLoginCrawler {
+	String ID;
```
변수 이름은 `static final` 형이 아니라면 소문자로 작성해주는게 좋을 것 같습니다.

juhyeonme commented on authentication-api/src/main/java/com/bluebird/pipit/portal/PortalLoginCrawler.java:
```diff
@@ -0,0 +1,65 @@
+
+	public Boolean loginPortal() throws Exception {
+
+		Connection.Response loginPageResponse = Jsoup.connect("https://portal.sookmyung.ac.kr/irj/portal")
+			.timeout(3000)
+			.header("Origin", "https://portal.sookmyung.ac.kr")
+			.header("Referer", "https://portal.sookmyung.ac.kr/irj/portal")
+			.header("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9")
+			.header("Content-Type", "application/x-www-form-urlencoded")
+			.header("Accept-Encoding", "gzip, deflate, br")
+			.header("Accept-Language", "ko-KR,ko;q=0.9,en-US;q=0.8,en;q=0.7")
+			.method(Connection.Method.GET)
```
여기도 value 쪽에 해당하는 값들을 다음 프로세스로 가져와서, 따로 관리해주는 편이 나을 것 같아요
- `application.yml` 파일에 등록 -> `@ConfigurationProperties` class -> 여기 클래스에서 `get`

juhyeonme commented on authentication-api/src/main/java/com/bluebird/pipit/portal/PortalLoginCrawler.java:
```diff
@@ -0,0 +1,65 @@
+
+public class PortalLoginCrawler {
+	String ID;
+	String password;
```
클래스변수는 private final 로 선언해주는 편이 좋고, lombok 의 `@Value` 를 사용한다면 모든 클래스 멤버들을 `private final` 형으로 만들 수 있습니다. (#4)

juhyeonme commented on authentication-api/src/main/java/com/bluebird/pipit/portal/PortalLoginCrawler.java:
```diff
@@ -0,0 +1,65 @@
+	String ID;
+	String password;
+
+	public PortalLoginCrawler(String ID, String password) {
```
클래스가 ID, password 값을 가져서 `상태를 유지` 하게 하는 형태는 state 를 담은 클래스를 여러개 생성해야하므로 낭비가 될 수 있을 것 같아요.
id, password 를 loginPortal() 의 매개변수로 선언해주고, 클래스는 상태를 갖지 않도록 작성해주는 편이 좋을 것 같습니다.

그럼 이 Constructor 는 삭제해도 될것 같은데요,
나중에 도움이 될까봐 남겨드리자면, 
Bean 의존성을 위해 이렇게 생성자 DI 를 작성하실 순간이 오게 될 것 같은데
그럴때는 lombok 의 @RequiredArgsConstructor 를 사용한다면 `모든 final 멤버들에 대하여 생성자 DI 자동 생성` 의 역할을 해줍니다

juhyeonme commented on authentication-api/src/main/java/com/bluebird/pipit/portal/PortalLoginCrawler.java:
```diff
@@ -0,0 +1,65 @@
+		this.password = password;
+	}
+
+	public Boolean loginPortal() throws Exception {
```
return 타입을 wrapper class 로 사용하신 이유가 있을까요?
wrapper 타입은 nullable 하기때문에 primitive 타입으로 변경해준다면 좋을 것 같습니다.

juhyeonme replied:
Exception 은 throw 로 던져버리지 않고, 범위를 좁혀서 (정확한 Exception class 명과 함께) 로깅과 함께 `catch` 해주는 편이 좋습니다.
예외가 전파되는것보다 해당 클래스에서 잡고 처리해주는 게 나을 것 같아요.

juhyeonme commented:
@kohseongyeon 
일단 1차적으로 적용할만한 부분들을 리뷰했습니다.
보완이 완료된다면 다시 request 주시면 2차 리뷰 진행할게여 😊

kohseongyeon committed:
1차 리뷰 반영

- jsoup 라이브러리 버전을 ext 정의하여 관리
- 불필요한 변수 정리
- JSESSIONID private static final 선언
- loginPortal() return 타입 변경
- Exception catch, logging
- Connection 설정 data를 ConnectionConfig 클래스로 관리

kohseongyeon commented on authentication-api/build.gradle:
```diff
@@ -24,8 +24,13 @@ configurations {
+
 dependencies {
     implementation("org.springframework.boot:spring-boot-starter-web")
+    implementation("org.jsoup:jsoup:${JSOUP_VERSION}")
```
jsoup 라이브러리는 spring dependency의 관리를 받지 못하여, ext로 버전을 따로 정의해주었습니다.

shindongsun0 commented on authentication-api/src/main/java/com/bluebird/pipit/portal/PortalLoginCrawler.java:
```diff
@@ -0,0 +1,78 @@
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2
```
로깅 프레임워크 전환이 용이한 slf4j 를 사용하는 것은 어떨까요?
slf4j를 사용함으로서 얻을 수 있는 이점은 기존에 사용하는 로깅 시스템(지금은 log4j2)을 만약 나중에 교체하고 싶을 때, 소스 코드를 수정하지 않고도 gradle의 의존성 설정만 바꾸면 손쉽게 적용할 수 있다는 점이 장점입니다.

juhyeonme commented on authentication-api/src/main/java/com/bluebird/pipit/portal/ConnectionConfig.java:
```diff
@@ -0,0 +1,18 @@
+
+@Component
+@ConfigurationProperties(prefix = "connection")
+@Getter @Setter
```
여기 `@Setter` 은 configuration properties 를 위한 애노테이션이지요?
원래는 지양해야 하는 키워드를 사용하고 있으니, 남용 가능성을 최소화하기 위해 Access Level 을 설정해주는 것도 좋은 방법입니다.

```java
@Setter(AccessLevel.PACKAGE)
```

그리고 getter 와 setter 애노테이션은 다른 줄에 선언해주는게 보기 좋을것 같아요

juhyeonme commented on authentication-api/src/main/java/com/bluebird/pipit/portal/ConnectionConfig.java:
```diff
@@ -0,0 +1,18 @@
+import lombok.Getter;
+import lombok.Setter;
+
+@Component
```
`@ConfigurationProperties` 만 붙여도 스프링 빈으로 생성될 수 있지 않나요?
`@Component` 는 생략해도 좋을것 같습니다

juhyeonme commented on authentication-api/src/main/java/com/bluebird/pipit/portal/ConnectionConfig.java:
```diff
@@ -0,0 +1,18 @@
+import lombok.Setter;
+
+@Component
+@ConfigurationProperties(prefix = "connection")
```
prefix 에 `connection` 뿐만 아니라, parent source 까지 표현해주면 읽기 더 쉬울것 같아요.
`portal.jsoup.connection` 과 같은 data path 표현이 있으면 좋을것 같습니다

juhyeonme commented on authentication-api/src/main/java/com/bluebird/pipit/portal/PortalLoginCrawler.java:
```diff
@@ -0,0 +1,78 @@
+@Log4j2
+@RequiredArgsConstructor
+public class PortalLoginCrawler {
+	private static final String JSESSIONID = "JSESSIONMARKID";
```
effective java 에서는 static 상수를 네이밍할때 언더바(`_`) 를 사용하라는 가이드를 내려주고 있습니다.
여기도 `JSESSION_ID` 로 쓴다면 더 읽기 쉬울것 같군요 ㅎㅎ

juhyeonme commented on authentication-api/src/main/java/com/bluebird/pipit/portal/PortalLoginCrawler.java:
```diff
@@ -0,0 +1,78 @@
+
+	public boolean loginPortal(String id, String password) {
+		Connection.Response loginPageResponse = null;
+		try {
+			loginPageResponse = Jsoup.connect(connectionConfig.getUrl())
+				.timeout(3000)
+				.header("Origin", connectionConfig.getHeader().get("origin"))
+				.header("Referer", connectionConfig.getHeader().get("referer"))
+				.header("Accept", connectionConfig.getHeader().get("accept"))
+				.header("Content-Type", connectionConfig.getHeader().get("contentType"))
+				.header("Accept-Encoding", connectionConfig.getHeader().get("acceptEncoding"))
+				.header("Accept-Language", connectionConfig.getHeader().get("acceptLanguage"))
+				.method(Connection.Method.GET)
+				.execute();
+		} catch (IOException e) {
+			log.error("Failed to connect to login page : GET", e);
+		}
+
+		Map<String, String> loginPageCookie = loginPageResponse.cookies();
```
여기서 loginPageResponse 셋팅이 실패하게 되면 `loginPageResponse` 은 `null` 로 남게되는건가요?
그럼 여기 `loginPageResponse.cookies()` 에서 NPE 를 던지게될 것 같습니다.

loginPageResponse 셋팅이 실패했을 때에는 아예 포털 인증 자체를 실패하게 만들고 에러메시지와 함께 (이건 적어주셨군요) 메소드를 return 시켜버리는 게 좋을 것 같습니다.

juhyeonme replied:
생각해보니 이건 화면에서도 포털인증이 실패함을 인지하는게 좋을것 같기도 해서..
일단은 return 으로 돌려보내고, 이후 PR 에서 정의된 exception 을 로그와함께 throw 해버리는것도 좋을것 같아요

juhyeonme commented on authentication-api/src/main/java/com/bluebird/pipit/portal/PortalLoginCrawler.java:
```diff
@@ -0,0 +1,78 @@
+		Map<String, String> loginPageCookie = loginPageResponse.cookies();
+
+		Document loginPageDocument = null;
+		try {
+			loginPageDocument = loginPageResponse.parse();
+		} catch (IOException e) {
+			log.error("Failed to parse response", e);
+		}
+		String j_salt = loginPageDocument.select("input.j_salt").val();
```
여기도 마찬가지로 loginPageDocument 이 try 구간에서 셋팅 실패했을때 NPE 가능성이 있습니다.
생각해두신 기획 의도대로 return 시키던가, 대체 데이터를 넣어주시는게 좋을것 같아요

juhyeonme commented on authentication-api/src/main/java/com/bluebird/pipit/portal/PortalLoginCrawler.java:
```diff
@@ -0,0 +1,78 @@
+		loginFormData.put("saveid", "N");
+
+		Connection.Response loginResponse = null;
+		try {
+			loginResponse = Jsoup.connect(connectionConfig.getUrl())
+				.userAgent(connectionConfig.getUserAgent())
+				.timeout(3000)
+				.header("Origin", connectionConfig.getHeader().get("origin"))
+				.header("Referer", connectionConfig.getHeader().get("referer"))
+				.header("Accept", connectionConfig.getHeader().get("accept"))
+				.header("Content-Type", connectionConfig.getHeader().get("contentType"))
+				.header("Accept-Encoding", connectionConfig.getHeader().get("acceptEncoding"))
+				.header("Accept-Language", connectionConfig.getHeader().get("acceptLanguage"))
+				.cookies(loginPageCookie)
+				.data(loginFormData)
+				.method(Connection.Method.POST)
+				.execute();
+		} catch (IOException e) {
+			log.error("Failed to connect to login page : POST", e);
+		}
+
+		return loginResponse.cookies().containsKey(JSESSIONID);
```
`loginResponse` 도 위와 같은 입장입니다. NPE 를 어떻게 방지하면 좋을지 고민해보면 좋을것 같아요 ㅎㅎ

juhyeonme commented on authentication-api/build.gradle:
```diff
@@ -24,8 +24,13 @@ configurations {
 }
 
+ext {
+    JSOUP_VERSION = "1.8.3"
```
👍🏻

juhyeonme commented:
리뷰 적용이 완료되면 여기 화살표를 눌러서 재리뷰 요청해주시면 될거같아요 ㅎㅎ
그 전까지는 작업중이라고 생각할수도 있으니, 표현해주면 좋을것 같습니다 

![스크린샷 2021-03-30 오후 4 03 09](https://user-images.githubusercontent.com/45280737/112947178-72e6ea80-9171-11eb-8254-c02dad12c22d.png)

juhyeonme commented on authentication-api/src/main/java/com/bluebird/pipit/portal/PortalLoginCrawler.java:
```diff
@@ -0,0 +1,78 @@
+		Connection.Response loginPageResponse = null;
+		try {
+			loginPageResponse = Jsoup.connect(connectionConfig.getUrl())
+				.timeout(3000)
```
뭔가 이런 값도 `private static final` 이 되면 좋을것 같군요

juhyeonme commented on authentication-api/src/main/java/com/bluebird/pipit/portal/ConnectionConfig.java:
```diff
@@ -0,0 +1,18 @@
+public class ConnectionConfig {
+	private String url;
+	private String userAgent;
+	private Map<String, String> header;
```
header 내부에 있는 애들을 `Map<String, String>` 말고 Object 로 선언할 수는 없을까요?
ConnectConfig 안에 `static class` 로 내부클래스 하나를 선언해서.. 뭔가 header 안에 애들도 존재감을 가지면 좋을것 같군요

kohseongyeon committed:
2차 리뷰 반영

- slf4j 사용
- `@Setter` Access Level 설정
- `@Component` 생략
- application.yml data path 표현
- ConnectionProperties Header 내부 클래스 선언
- NPE 방지

kohseongyeon commented:
2차 리뷰까지 반영했습니다! 꼼꼼하고 자세한 리뷰 너무 감사드려요..❤

juhyeonme reviewed (approved):
굿굿 리뷰를 아주 완벽하게 반영해줬네여😍 
동선님 리뷰 후 머지하고 다음 테스크 진행하셔도 될것 같습니다. 👍🏻 

최종 커밋 이름은 `[#19] 숙명포털 로그인 crawler 작성 (#20)` 이라고 적어주는게 좋을 것 같아요

juhyeonme commented on authentication-api/src/main/java/com/bluebird/pipit/portal/PortalLoginCrawler.java:
```diff
@@ -0,0 +1,82 @@
+@RequiredArgsConstructor
+public class PortalLoginCrawler {
+	private static final String JSESSION_ID = "JSESSIONMARKID";
+	private static final int TIMEOUT = 3000;
```
이 `TIMEOUT` 이 connection timeout 인가요?
그렇다면 이름도 `CONNECTION_TIMEOUT` 으로 작성해주는게 더 좋을것 같군요 ㅎㅎ

juhyeonme commented:
위에 드린 `timeout` 관련 리뷰는 반영 후 리뷰 재요청 없이 바로 merge 하면 될것 같습니당

shindongsun0 reviewed (approved):
주현님 리뷰 반영 후 다음 테스크 저랑 함께 하시져~ 👍

kohseongyeon committed:
TIMEOUT -> CONNECTION_TIMEOUT 이름 변경
